### ISSUE TITLE
Scylla 5.0.9 segfault

### ISSUE THREAD

**Issue by santithorn** [scylladb/scylladb#12632]
This is Scylla's bug tracker, to be used for reporting bugs only.
If you have a question about Scylla, and not a bug, please ask it in
our mailing-list at scylladb-dev@googlegroups.com or in our slack channel.

- [v] I have read the disclaimer above, and I am reporting a suspected malfunction in Scylla.

*Installation details*
Scylla version (or git commit hash): 5.0.9
Cluster size: 3
OS (RHEL/CentOS/Ubuntu/AWS AMI): CentOS 7

*Hardware details (for performance issues)*          Delete if unneeded
Platform (physical/VM/cloud instance type/docker): VM (KVM)
Hardware: vcores=4 memory=32GB
Disks: (SSD/HDD, count) HDD

health check report: 08d974ea-b238-4845-9ac0-30c80f4b997d/prod-data2-health-check-report.txt
core dump: 08d974ea-b238-4845-9ac0-30c80f4b997d/core.scylla.995.05f478d8544d4ee08fa36934d05bb33a.1228.1674525348000000.xz

Scylla 5.0.9 crashed with core dump, don't know what trigger this. Please help check.

```
Jan 24 08:55:48 prod-data2 scylla: Segmentation fault on shard 2.
Jan 24 08:55:48 prod-data2 scylla: Backtrace:
Jan 24 08:55:48 prod-data2 scylla: 0x4445718
Jan 24 08:55:48 prod-data2 scylla: 0x44761f6
Jan 24 08:55:48 prod-data2 scylla: 0x4ac5fd1
Jan 24 08:55:48 prod-data2 scylla: 0x7f8d6c8e5a1f
Jan 24 08:55:48 prod-data2 scylla: 0x447b5b3
Jan 24 08:55:48 prod-data2 scylla: 0x24c422b
Jan 24 08:55:48 prod-data2 scylla: 0x2b715cb
Jan 24 08:55:48 prod-data2 scylla: 0x4456674
Jan 24 08:55:48 prod-data2 scylla: 0x4457a47
Jan 24 08:55:48 prod-data2 scylla: 0x4476765
Jan 24 08:55:48 prod-data2 scylla: 0x442a4ca
Jan 24 08:55:48 prod-data2 scylla: /opt/scylladb/libreloc/libpthread.so.0+0x9298
Jan 24 08:55:48 prod-data2 scylla: /opt/scylladb/libreloc/libc.so.6+0x100352
Jan 24 08:55:48 prod-data2 kernel: reactor-2[1230]: segfault at 8a ip 000000000447b5b4 sp 00007f8d691d3480 error 4 in scylla[fa4000+3e23000]
Jan 24 08:56:53 prod-data2 systemd-coredump: Process 1228 (scylla) of user 995 dumped core.#012#012Stack trace of thread 1230:#012#0  0x000000000447b5b4 _ZN7seastar5timerINS_12lowres_clockEED1Ev (scylla)#012#1  0x00000000024c422c _ZN7seastar15basic_semaphoreINS_35semaphore_default_exception_factoryENS_12lowres_clockEE4waitENSt6chrono10time_pointIS2_NS4_8durationIlSt5ratioILl1ELl1000000000EEEEEEm (scylla)#012#2  0x0000000002b715cc _ZN7seastar12continuationINS_8internal22promise_base_with_typeIN10row_locker11lock_holderEEEZNS3_7lock_ckERKN3dht13decorated_keyERK21clustering_key_prefixbNSt6chrono10time_pointINS_12lowres_clockENSD_8durationIlSt5ratioILl1ELl1000000000EEEEEERNS3_5statsEE3$_2ZNS_6futureINS_15semaphore_unitsINS_35semaphore_default_exception_factoryESF_EEE14then_impl_nrvoISN_NSO_IS4_EEEET0_OT_EUlOS5_RSN_ONS_12future_stateISR_EEE_SR_E15run_and_disposeEv (scylla)#012#3  0x0000000004456675 _ZN7seastar7reactor14run_some_tasksEv (scylla)#012#4  0x0000000004457a48 _ZN7seastar7reactor6do_runEv (scylla)#012#5  0x0000000004476766 _ZNSt17_Function_handlerIFvvEZN7seastar3smp9configureERKNS1_11smp_optionsERKNS1_15reactor_optionsEE5$_100E9_M_invokeERKSt9_Any_data (scylla)#012#6  0x000000000442a4cb _ZN7seastar12posix_thread13start_routineEPv (scylla)#012#7  0x00007f8d6c8db299 start_thread (libpthread.so.0)#012#8  0x00007f8d6bded353 __clone (libc.so.6)#012#012Stack trace of thread 1233:#012#0  0x00007f8d6bde25bf __poll (libc.so.6)#012#1  0x000000000449b6c6 _ZN7seastar21reactor_backend_epoll26task_quota_timer_thread_fnEv (scylla)#012#2  0x00007f8d6bfb0c84 execute_native_thread_routine (libstdc++.so.6)#012#3  0x00007f8d6c8db299 start_thread (libpthread.so.0)#012#4  0x00007f8d6bded353 __clone (libc.so.6)#012#012Stack trace of thread 1236:#012#0  0x00007f8d6c8e494c read (libpthread.so.0)#012#1  0x0000000004498ba5 _ZN7seastar11thread_pool4workENS_13basic_sstringIcjLj15ELb1EEE (scylla)#012#2  0x0000000004498f10 _ZNSt17_Function_handlerIFvvEZN7seastar11thread_poolC1EPNS1_7reactorENS1_13basic_sstringIcjLj15ELb1EEEE3$_0E9_M_invokeERKSt9_Any_data (scylla)#012#3  0x000000000442a4cb _ZN7seastar12posix_thread13start_routineEPv (scylla)#012#4  0x00007f8d6c8db299 start_thread (libpthread.so.0)#012#5  0x00007f8d6bded353 __clone (libc.so.6)#012#012Stack trace of thread 1239:#012#0  0x00007f8d6bde25bf __poll (libc.so.6)#012#1  0x000000000449b6c6 _ZN7seastar21reactor_backend_epoll26task_quota_timer_thread_fnEv (scylla)#012#2  0x00007f8d6bfb0c84 execute_native_thread_routine (libstdc++.so.6)#012#3  0x00007f8d6c8db299 start_thread (libpthread.so.0)#012#4  0x00007f8d6bded353 __clone (libc.so.6)#012#012Stack trace of thread 1232:#012#0  0x00007f8d6c8e494c read (libpthread.so.0)#012#1  0x0000000004498ba5 _ZN7seastar11thread_pool4workENS_13basic_sstringIcjLj15ELb1EEE (scylla)#012#2  0x0000000004498f10 _ZNSt17_Function_handlerIFvvEZN7seastar11thread_poolC1EPNS1_7reactorENS1_13basic_sstringIcjLj15ELb1EEEE3$_0E9_M_invokeERKSt9_Any_data (scylla)#012#3  0x000000000442a4cb _ZN7seastar12posix_thread13start_routineEPv (scylla)#012#4  0x00007f8d6c8db299 start_thread (libpthread.so.0)#012#5  0x00007f8d6bded353 __clone (libc.so.6)#012#012Stack trace of thread 1228:#012#0  0x000000000236b66d _ZN7service13storage_proxy10mutate_endEN7seastar6futureIvEEN5utils15latency_counterERNS_19storage_proxy_stats11write_statsEN7tracing15trace_state_ptrE (scylla)#012#1  0x000000000242ad90 _ZN7seastar12continuationINS_8internal22promise_base_with_typeIvEEZN7service13storage_proxy15mutate_internalIN5boost14iterator_rangeIN9__gnu_cxx17__normal_iteratorIP8mutationSt6vectorISB_SaISB_EEEEEEEENS_6futureIvEET_N2db17consistency_levelEbN7tracing15trace_state_ptrE14service_permitSt8optionalINSt6chrono10time_pointINS_12lowres_clockENSR_8durationIlSt5ratioILl1ELl1000000000EEEEEEENS_13lw_shared_ptrIN3cdc24operation_result_trackerEEEEUlSJ_E_ZNSJ_17then_wrapped_nrvoISJ_S14_EENS_8futurizeISK_E4typeEOT0_EUlOS3_RS14_ONS_12future_stateINS1_9monostateEEEE_vE15run_and_disposeEv (scylla)#012#2  0x0000000004456675 _ZN7seastar7reactor14run_some_tasksEv (scylla)#012#3  0x0000000004457a48 _ZN7seastar7reactor6do_runEv (scylla)#012#4  0x0000000004456cad _ZN7seastar7reactor3runEv (scylla)#012#5  0x00000000043fd039 _ZN7seastar12app_template14run_deprecatedEiPPcOSt8functionIFvvEE (scylla)#012#6  0x00000000043fc517 _ZN7seastar12app_template3runEiPPcOSt8functionIFNS_6futureIiEEvEE (scylla)#012#7  0x0000000000fa78fe _ZL11scylla_mainiPPc (scylla)#012#8  0x0000000000fa51db main (scylla)#012#9  0x00007f8d6bd14b75 __libc_start_main (libc.so.6)#012#10 0x0000000000fa41ae _start (scylla)#012#012Stack trace of thread 1238:#012#0  0x00007f8d6c8e494c read (libpthread.so.0)#012#1  0x0000000004498ba5 _ZN7seastar11thread_pool4workENS_13basic_sstringIcjLj15ELb1EEE (scylla)#012#2  0x0000000004498f10 _ZNSt17_Function_handlerIFvvEZN7seastar11thread_poolC1EPNS1_7reactorENS1_13basic_sstringIcjLj15ELb1EEEE3$_0E9_M_invokeERKSt9_Any_data (scylla)#012#3  0x000000000442a4cb _ZN7seastar12posix_thread13start_routineEPv (scylla)#012#4  0x00007f8d6c8db299 start_thread (libpthread.so.0)#012#5  0x00007f8d6bded353 __clone (libc.so.6)#012#012Stack trace of thread 1237:#012#0  0x00007f8d6bde25bf __poll (libc.so.6)#012#1  0x000000000449b6c6 _ZN7seastar21reactor_backend_epoll26task_quota_timer_thread_fnEv (scylla)#012#2  0x00007f8d6bfb0c84 execute_native_thread_routine (libstdc++.so.6)#012#3  0x00007f8d6c8db299 start_thread (libpthread.so.0)#012#4  0x00007f8d6bded353 __clone (libc.so.6)#012#012Stack trace of thread 1231:#012#0  0x00000000013e73be _ZN11intrusive_b4nodeI10rows_entryXadL_ZNS1_5_linkEEENS1_11tri_compareELm12ELm20ELNS_10key_searchE0ELNS_10with_debugE0EE10remove_keyEm (scylla)#012#1  0x0000000001403e69 _ZN10rows_entry10on_evictedER13cache_tracker (scylla)#012#2  0x000000000143a9b3 _ZNSt17_Function_handlerIFN7seastar6memory17reclaiming_resultEvEZN13cache_trackerC1ER26mutation_application_statsNS0_10bool_classI20register_metrics_tagEEE3$_0E9_M_invokeERKSt9_Any_data (scylla)#012#3  0x00000000014d15ed _ZN8logalloc11region_impl10evict_someEv (scylla)#012#4  0x00000000014dd0e9 _ZN8logallocL22reclaim_from_evictableERNS_11region_implEmN7seastar10bool_classI18is_preemptible_tagEE (scylla)#012#5  0x00000000014d8dc8 _ZN8logalloc7tracker4impl24compact_and_evict_lockedEmmN7seastar10bool_classI18is_preemptible_tagEE (scylla)#012#6  0x00000000014c9c27 _ZN8logalloc7tracker4impl17compact_and_evictEmmN7seastar10bool_classI18is_preemptible_tagEE (scylla)#012#7  0x00000000014c8560 _ZN8logalloc12segment_pool16allocate_segmentEm (scylla)#012#8  0x00000000014e6c58 _ZN8logalloc18allocating_section7reserveEv (scylla)#012#9  0x00000000018aec2d _ZN8sstables27index_consume_entry_contextINS_14index_consumerEE13process_stateERN7seastar16temporary_bufferIcEE (scylla)#012#10 0x00000000018ab75f _ZZN7seastar12input_streamIcE7consumeISt17reference_wrapperIN8sstables27index_consume_entry_contextINS4_14index_consumerEEEEEENS_6futureIvEEOT_ENUlvE_clEv (scylla)#012#11 0x00000000018b4ce1 _ZN7seastar8internal8repeaterIZNS_12input_streamIcE7consumeISt17reference_wrapperIN8sstables27index_consume_entry_contextINS6_14index_consumerEEEEEENS_6futureIvEEOT_EUlvE_E15run_and_disposeEv (scylla)#012#12 0x0000000004456675 _ZN7seastar7reactor14run_some_tasksEv (scylla)#012#13 0x0000000004457a48 _ZN7seastar7reactor6do_runEv (scylla)#012#14 0x0000000004476766 _ZNSt17_Function_handlerIFvvEZN7seastar3smp9configureERKNS1_11smp_optionsERKNS1_15reactor_optionsEE5$_100E9_M_invokeERKSt9_Any_data (scylla)#012#15 0x000000000442a4cb _ZN7seastar12posix_thread13start_routineEPv (scylla)#012#16 0x00007f8d6c8db299 start_thread (libpthread.so.0)#012#17 0x00007f8d6bded353 __clone (libc.so.6)#012#012Stack trace of thread 1235:#012#0  0x00007f8d6bde25bf __poll (libc.so.6)#012#1  0x000000000449b6c6 _ZN7seastar21reactor_backend_epoll26task_quota_timer_thread_fnEv (scylla)#012#2  0x00007f8d6bfb0c84 execute_native_thread_routine (libstdc++.so.6)#012#3  0x00007f8d6c8db299 start_thread (libpthread.so.0)#012#4  0x00007f8d6bded353 __clone (libc.so.6)#012#012Stack trace of thread 1234:#012#0  0x00007f8d6c8e494c read (libpthread.so.0)#012#1  0x0000000004498ba5 _ZN7seastar11thread_p
```

**Comment by avikivity** [comment]
```
Jan 24 08:55:48 prod-data2 scylla: Backtrace:Jan 24 08:55:48 prod-data2 scylla:
[Backtrace #0]
void seastar::backtrace<seastar::backtrace_buffer::append_backtrace()::{lambda(seastar::frame)#1}>(seastar::backtrace_buffer::append_backtrace()::{lambda(seastar::frame)#1}&&) at ./build/release/seastar/./seastar/include/seastar/util/backtrace.hh:59
 (inlined by) seastar::backtrace_buffer::append_backtrace() at ./build/release/seastar/./seastar/src/core/reactor.cc:750
 (inlined by) seastar::print_with_backtrace(seastar::backtrace_buffer&, bool) at ./build/release/seastar/./seastar/src/core/reactor.cc:780
seastar::print_with_backtrace(char const*, bool) at ./build/release/seastar/./seastar/src/core/reactor.cc:792
 (inlined by) seastar::sigsegv_action() at ./build/release/seastar/./seastar/src/core/reactor.cc:3668
 (inlined by) operator() at ./build/release/seastar/./seastar/src/core/reactor.cc:3654
 (inlined by) __invoke at ./build/release/seastar/./seastar/src/core/reactor.cc:3650
wasmtime_runtime::traphandlers::unix::trap_handler at wasmtime_runtime.9xv9ixev-cgu.15:?
?? ??:0
~timer at ./build/release/seastar/./seastar/src/core/reactor.cc:579
seastar::basic_semaphore<seastar::semaphore_default_exception_factory, seastar::lowres_clock>::wait(std::chrono::time_point<seastar::lowres_clock, std::chrono::duration<long, std::ratio<1l, 1000000000l> > >, unsigned long) at paxos_state.cc:?
 (inlined by) seastar::basic_semaphore<seastar::semaphore_default_exception_factory, seastar::lowres_clock>::wait(std::chrono::time_point<seastar::lowres_clock, std::chrono::duration<long, std::ratio<1l, 1000000000l> > >, unsigned long) at ././seastar/include/seastar/core/semaphore.hh:197
seastar::future<seastar::semaphore_units<seastar::semaphore_default_exception_factory, seastar::lowres_clock> > seastar::get_units<seastar::semaphore_default_exception_factory, seastar::lowres_clock>(seastar::basic_semaphore<seastar::semaphore_default_exception_factory, seastar::lowres_clock>&, unsigned long, seastar::basic_semaphore<seastar::semaphore_default_exception_factory, seastar::lowres_clock>::time_point) at ././seastar/include/seastar/core/semaphore.hh:455
 (inlined by) seastar::basic_rwlock<seastar::lowres_clock>::hold_write_lock(std::chrono::time_point<seastar::lowres_clock, std::chrono::duration<long, std::ratio<1l, 1000000000l> > >) at ././seastar/include/seastar/core/rwlock.hh:164
 (inlined by) operator()<seastar::semaphore_units<seastar::semaphore_default_exception_factory, seastar::lowres_clock> > at ./db/view/row_locking.cc:104
 (inlined by) seastar::future<row_locker::lock_holder> std::__invoke_impl<seastar::future<row_locker::lock_holder>, row_locker::lock_ck(dht::decorated_key const&, clustering_key_prefix const&, bool, std::chrono::time_point<seastar::lowres_clock, std::chrono::duration<long, std::ratio<1l, 1000000000l> > >, row_locker::stats&)::$_2&, seastar::semaphore_units<seastar::semaphore_default_exception_factory, seastar::lowres_clock> >(std::__invoke_other, row_locker::lock_ck(dht::decorated_key const&, clustering_key_prefix const&, bool, std::chrono::time_point<seastar::lowres_clock, std::chrono::duration<long, std::ratio<1l, 1000000000l> > >, row_locker::stats&)::$_2&, seastar::semaphore_units<seastar::semaphore_default_exception_factory, seastar::lowres_clock>&&) at /usr/lib/gcc/x86_64-redhat-linux/11/../../../../include/c++/11/bits/invoke.h:61
 (inlined by) std::__invoke_result<row_locker::lock_ck(dht::decorated_key const&, clustering_key_prefix const&, bool, std::chrono::time_point<seastar::lowres_clock, std::chrono::duration<long, std::ratio<1l, 1000000000l> > >, row_locker::stats&)::$_2&, seastar::semaphore_units<seastar::semaphore_default_exception_factory, seastar::lowres_clock> >::type std::__invoke<row_locker::lock_ck(dht::decorated_key const&, clustering_key_prefix const&, bool, std::chrono::time_point<seastar::lowres_clock, std::chrono::duration<long, std::ratio<1l, 1000000000l> > >, row_locker::stats&)::$_2&, seastar::semaphore_units<seastar::semaphore_default_exception_factory, seastar::lowres_clock> >(row_locker::lock_ck(dht::decorated_key const&, clustering_key_prefix const&, bool, std::chrono::time_point<seastar::lowres_clock, std::chrono::duration<long, std::ratio<1l, 1000000000l> > >, row_locker::stats&)::$_2&, seastar::semaphore_units<seastar::semaphore_default_exception_factory, seastar::lowres_clock>&&) at /usr/lib/gcc/x86_64-redhat-linux/11/../../../../include/c++/11/bits/invoke.h:96
 (inlined by) std::invoke_result<row_locker::lock_ck(dht::decorated_key const&, clustering_key_prefix const&, bool, std::chrono::time_point<seastar::lowres_clock, std::chrono::duration<long, std::ratio<1l, 1000000000l> > >, row_locker::stats&)::$_2&, seastar::semaphore_units<seastar::semaphore_default_exception_factory, seastar::lowres_clock> >::type std::invoke<row_locker::lock_ck(dht::decorated_key const&, clustering_key_prefix const&, bool, std::chrono::time_point<seastar::lowres_clock, std::chrono::duration<long, std::ratio<1l, 1000000000l> > >, row_locker::stats&)::$_2&, seastar::semaphore_units<seastar::semaphore_default_exception_factory, seastar::lowres_clock> >(row_locker::lock_ck(dht::decorated_key const&, clustering_key_prefix const&, bool, std::chrono::time_point<seastar::lowres_clock, std::chrono::duration<long, std::ratio<1l, 1000000000l> > >, row_locker::stats&)::$_2&, seastar::semaphore_units<seastar::semaphore_default_exception_factory, seastar::lowres_clock>&&) at /usr/lib/gcc/x86_64-redhat-linux/11/../../../../include/c++/11/functional:97
 (inlined by) auto seastar::internal::future_invoke<row_locker::lock_ck(dht::decorated_key const&, clustering_key_prefix const&, bool, std::chrono::time_point<seastar::lowres_clock, std::chrono::duration<long, std::ratio<1l, 1000000000l> > >, row_locker::stats&)::$_2&, seastar::semaphore_units<seastar::semaphore_default_exception_factory, seastar::lowres_clock> >(row_locker::lock_ck(dht::decorated_key const&, clustering_key_prefix const&, bool, std::chrono::time_point<seastar::lowres_clock, std::chrono::duration<long, std::ratio<1l, 1000000000l> > >, row_locker::stats&)::$_2&, seastar::semaphore_units<seastar::semaphore_default_exception_factory, seastar::lowres_clock>&&) at ././seastar/include/seastar/core/future.hh:1211
 (inlined by) operator() at ././seastar/include/seastar/core/future.hh:1582
 (inlined by) void seastar::futurize<seastar::future<row_locker::lock_holder> >::satisfy_with_result_of<seastar::future<seastar::semaphore_units<seastar::semaphore_default_exception_factory, seastar::lowres_clock> >::then_impl_nrvo<row_locker::lock_ck(dht::decorated_key const&, clustering_key_prefix const&, bool, std::chrono::time_point<seastar::lowres_clock, std::chrono::duration<long, std::ratio<1l, 1000000000l> > >, row_locker::stats&)::$_2, seastar::future<row_locker::lock_holder> >(row_locker::lock_ck(dht::decorated_key const&, clustering_key_prefix const&, bool, std::chrono::time_point<seastar::lowres_clock, std::chrono::duration<long, std::ratio<1l, 1000000000l> > >, row_locker::stats&)::$_2&&)::{lambda(seastar::internal::promise_base_with_type<row_locker::lock_holder>&&, row_locker::lock_ck(dht::decorated_key const&, clustering_key_prefix const&, bool, std::chrono::time_point<seastar::lowres_clock, std::chrono::duration<long, std::ratio<1l, 1000000000l> > >, row_locker::stats&)::$_2&, seastar::future_state<seastar::semaphore_units<seastar::semaphore_default_exception_factory, seastar::lowres_clock> >&&)#1}::operator()(seastar::internal::promise_base_with_type<row_locker::lock_holder>&&, row_locker::lock_ck(dht::decorated_key const&, clustering_key_prefix const&, bool, std::chrono::time_point<seastar::lowres_clock, std::chrono::duration<long, std::ratio<1l, 1000000000l> > >, row_locker::stats&)::$_2&, seastar::future_state<seastar::semaphore_units<seastar::semaphore_default_exception_factory, seastar::lowres_clock> >&&) const::{lambda()#1}>(seastar::internal::promise_base_with_type<row_locker::lock_holder>&&, row_locker::lock_ck(dht::decorated_key const&, clustering_key_prefix const&, bool, std::chrono::time_point<seastar::lowres_clock, std::chrono::duration<long, std::ratio<1l, 1000000000l> > >, row_locker::stats&)::$_2&&) at ././seastar/include/seastar/core/future.hh:2120
 (inlined by) operator() at ././seastar/include/seastar/core/future.hh:1575
 (inlined by) seastar::continuation<seastar::internal::promise_base_with_type<row_locker::lock_holder>, row_locker::lock_ck(dht::decorated_key const&, clustering_key_prefix const&, bool, std::chrono::time_point<seastar::lowres_clock, std::chrono::duration<long, std::ratio<1l, 1000000000l> > >, row_locker::stats&)::$_2, seastar::future<seastar::semaphore_units<seastar::semaphore_default_exception_factory, seastar::lowres_clock> >::then_impl_nrvo<row_locker::lock_ck(dht::decorated_key const&, clustering_key_prefix const&, bool, std::chrono::time_point<seastar::lowres_clock, std::chrono::duration<long, std::ratio<1l, 1000000000l> > >, row_locker::stats&)::$_2, seastar::future<row_locker::lock_holder> >(row_locker::lock_ck(dht::decorated_key const&, clustering_key_prefix const&, bool, std::chrono::time_point<seastar::lowres_clock, std::chrono::duration<long, std::ratio<1l, 1000000000l> > >, row_locker::stats&)::$_2&&)::{lambda(seastar::internal::promise_base_with_type<row_locker::lock_holder>&&, row_locker::lock_ck(dht::decorated_key const&, clustering_key_prefix const&, bool, std::chrono::time_point<seastar::lowres_clock, std::chrono::duration<long, std::ratio<1l, 1000000000l> > >, row_locker::stats&)::$_2&, seastar::future_state<seastar::semaphore_units<seastar::semaphore_default_exception_factory, seastar::lowres_clock> >&&)#1}, seastar::semaphore_units<seastar::semaphore_default_exception_factory, seastar::lowres_clock> >::run_and_dispose() at ././seastar/include/seastar/core/future.hh:767
seastar::reactor::run_tasks(seastar::reactor::task_queue&) at ./build/release/seastar/./seastar/src/core/reactor.cc:2348
 (inlined by) seastar::reactor::run_some_tasks() at ./build/release/seastar/./seastar/src/core/reactor.cc:2758
seastar::reactor::do_run() at ./build/release/seastar/./seastar/src/core/reactor.cc:2927
operator() at ./build/release/seastar/./seastar/src/core/reactor.cc:4137
 (inlined by) void std::__invoke_impl<void, seastar::smp::configure(seastar::smp_options const&, seastar::reactor_options const&)::$_100&>(std::__invoke_other, seastar::smp::configure(seastar::smp_options const&, seastar::reactor_options const&)::$_100&) at /usr/lib/gcc/x86_64-redhat-linux/11/../../../../include/c++/11/bits/invoke.h:61
 (inlined by) std::enable_if<is_invocable_r_v<void, seastar::smp::configure(seastar::smp_options const&, seastar::reactor_options const&)::$_100&>, void>::type std::__invoke_r<void, seastar::smp::configure(seastar::smp_options const&, seastar::reactor_options const&)::$_100&>(seastar::smp::configure(seastar::smp_options const&, seastar::reactor_options const&)::$_100&) at /usr/lib/gcc/x86_64-redhat-linux/11/../../../../include/c++/11/bits/invoke.h:111
 (inlined by) std::_Function_handler<void (), seastar::smp::configure(seastar::smp_options const&, seastar::reactor_options const&)::$_100>::_M_invoke(std::_Any_data const&) at /usr/lib/gcc/x86_64-redhat-linux/11/../../../../include/c++/11/bits/std_function.h:291
std::function<void ()>::operator()() const at /usr/lib/gcc/x86_64-redhat-linux/11/../../../../include/c++/11/bits/std_function.h:560
 (inlined by) seastar::posix_thread::start_routine(void*) at ./build/release/seastar/./seastar/src/core/posix.cc:60
/data/scylla-s3-reloc.cache/by-build-id/520004ed6991c6969da975000ec4a2ea10defe99/extracted/scylla/libreloc/libpthread.so.0: ELF 64-bit LSB shared object, x86-64, version 1 (SYSV), dynamically linked, interpreter /lib64/ld-linux-x86-64.so.2, BuildID[sha1]=219a7f6a92210920b2b6c86ef84d842b33324b0f, for GNU/Linux 3.2.0, not stripped

start_thread at ??:?
/data/scylla-s3-reloc.cache/by-build-id/520004ed6991c6969da975000ec4a2ea10defe99/extracted/scylla/libreloc/libc.so.6: ELF 64-bit LSB shared object, x86-64, version 1 (GNU/Linux), dynamically linked, interpreter /lib64/ld-linux-x86-64.so.2, BuildID[sha1]=9de0f6e78ec16db4c6c167efc6b37466705edeff, for GNU/Linux 3.2.0, not stripped
```

**Comment by avikivity** [comment]
Smells familiar.

**Comment by avikivity** [comment]
Duplicate of #12168.

**Comment by avikivity** [comment]
#12168 was fixed in 5.0.8, but this was reported on 5.0.9.

**Comment by avikivity** [comment]
@bhalevy ideas?

**Comment by bhalevy** [comment]
This wait call at paxos_state.cc looks weird.
will look into it tomorrow.

**Comment by bhalevy** [comment]
It looks like we were too careful not to make a radical fix in 5007ded2c19303ce1bc95cb6b313597b1b9de2d6 so to reduce the churn and risk, but I think we're still exposed to races.
The row lock may be constructed in parallel to read-locking the partition, so it might still deleted before used :-(

partition is locked in the background in https://github.com/scylladb/scylladb/blob/ce96b472d39a6c72f99176c8cf84e171143ae046/db/view/row_locking.cc#L89

row lock may be constructed in https://github.com/scylladb/scylladb/blob/ce96b472d39a6c72f99176c8cf84e171143ae046/db/view/row_locking.cc#L94

and until the lock_partition future is resolved in https://github.com/scylladb/scylladb/blob/ce96b472d39a6c72f99176c8cf84e171143ae046/db/view/row_locking.cc#L110-L111
the row lock in `j->second` might have been deleted.

**Comment by bhalevy** [comment]
The race is apparently with https://github.com/scylladb/scylladb/blob/ce96b472d39a6c72f99176c8cf84e171143ae046/db/view/row_locking.cc#L179
that erases the row-lock even if the partition lock might be held.

We can fix it by moving the search-for or allocation of the row lock to the continuation of lock_partition (which adds an allocation for copying the clustering key prefix) and/or
fixing the unlock code to erase row locks only when the partition is not locked.
The latter will likely leave behind unlocked row-locks, to be reclaimed only when the load subsides.
I'm not sure that's desired so I'll leave it for a potential follow-up and proceed with the former (serializing searching or creating the row lock entry after locking the partition (and starting the read/write lock for it synchronously so to keep it alive)

**Comment by michoecho** [comment]
Bumping for backport.

**Comment by avikivity** [comment]
Backported to 5.0, 5.1, 5.2.
